File: it-agents/skills/code-architect/references/code-examples/python/proxy.py
"""Proxy — a placeholder controlling access to a real object (lazy + caching)."""

from __future__ import annotations

from abc import ABC, abstractmethod


class Resource(ABC):
    @abstractmethod
    def fetch(self, key: str) -> str:
        ...


class RealResource(Resource):
    """Expensive to create and to query."""

    def __init__(self) -> None:
        print("[RealResource] expensive initialization")

    def fetch(self, key: str) -> str:
        return f"value-for-{key}"


class ResourceProxy(Resource):
    """Defers creation until first use and caches results."""

    def __init__(self) -> None:
        self._real: RealResource | None = None
        self._cache: dict[str, str] = {}

    def fetch(self, key: str) -> str:
        if key in self._cache:
            return f"{self._cache[key]} (cached)"
        if self._real is None:
            self._real = RealResource()
        value = self._real.fetch(key)
        self._cache[key] = value
        return value


if __name__ == "__main__":
    proxy = ResourceProxy()
    print(proxy.fetch("alpha"))
    print(proxy.fetch("alpha"))

File: it-agents/skills/code-architect/references/code-examples/cpp/Proxy.cpp
// Proxy — a placeholder controlling access to a real object (lazy/caching).
#include <iostream>
#include <memory>
#include <string>

// Common interface.
class Resource {
public:
    virtual ~Resource() = default;
    virtual std::string fetch() = 0;
};

// Expensive real subject.
class HeavyResource : public Resource {
    std::string payload;
public:
    HeavyResource() {
        std::cout << "(building heavy resource...)\n";
        payload = "heavy-payload";
    }
    std::string fetch() override { return payload; }
};

// Proxy: lazy construction + caching.
class ResourceProxy : public Resource {
    std::unique_ptr<HeavyResource> real;
    std::string cache;
    bool cached = false;
public:
    std::string fetch() override {
        if (!real) real = std::make_unique<HeavyResource>();
        if (!cached) {
            cache = real->fetch();
            cached = true;
        } else {
            std::cout << "(served from cache)\n";
        }
        return cache;
    }
};

int main() {
    ResourceProxy proxy;
    std::cout << "before first access\n";
    std::cout << proxy.fetch() << "\n";
    std::cout << proxy.fetch() << "\n";
    return 0;
}
